src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java

```diff
@@ -11,47 +11,79 @@
-                .transactionRetryPolicy(RetryPolicy.builder().maxRetries(3).build())
-                .sessionClientBuilder(QldbSessionClient.builder())
-                .build();
+        try {
+            qldbDriver = QldbDriver.builder()
+                    .ledger(config.getString("qldb_ledger_name"))
+                    .transactionRetryPolicy(RetryPolicy.builder().maxRetries(3).build())
+                    .sessionClientBuilder(QldbSessionClient.builder())
+                    .build();
+        }
+        catch (Exception e){
+            logger.error("cannot establish connection with QLDB");
+        }
         logTable = config.getString("qldb_table_name");
         qldbLogging = config.getBoolean("enable_qldb_admin_logging");
     }
     @Override
-    public void writeLog(AuditModel model) {
-        if(model == null){ //should never be true, but check exists in case
-            return;
-        }
-        if(qldbLogging) {
-            qldbDriver.execute(txn -> {
-                List<IonValue> sanitizedInputs = new ArrayList<>();
-                JsonObject jsonObject = model.writeToJson();
-                StringBuilder query = new StringBuilder("UPDATE " + logTable + " AS t SET data = ?");
-                sanitizedInputs.add(ionSys.newLoader().load(jsonObject.toString()).get(0));
-                query.append(" WHERE t.data.itemType = ? AND t.data.itemKey = ?");
-                sanitizedInputs.add(ionSys.newString(jsonObject.getString("itemType")));
-                sanitizedInputs.add(ionSys.newString(jsonObject.getString("itemKey")));
-
-                Result r = txn.execute(query.toString(), sanitizedInputs);
-                if (!r.iterator().hasNext()) {
-                    logger.warn("Malformed audit log input: no log written to QLDB");
+    public boolean writeLogs(Collection<AuditModel> models) {
+        AtomicBoolean successfulLog = new AtomicBoolean(true);
+        try {
+            if (qldbLogging) {
+                qldbDriver.execute(txn -> {
+                    for(AuditModel model : models){
+                        if(!(model instanceof QLDBAuditModel)){ //should never be true, but check in case
+                            successfulLog.set(false);
+                            logger.error("Only QLDBAuditModel should be passed into QLDBAuditWriter");
+                            txn.abort();
+                            break;
+                        }
+                        QLDBAuditModel qldbModel = (QLDBAuditModel) model;
+                        JsonObject jsonObject = qldbModel.writeToJson();
+                        String query;
+                        List<IonValue> sanitizedInputs = new ArrayList<>();
+                        if(qldbModel.actionTaken == Actions.CREATE){
+                            query = "INSERT INTO " + logTable + " VALUE ?";
+                            JsonObject wrapped = new JsonObject().put("data", jsonObject);
```

Comment:
are you sure this puts itemType and itemKey at top level? Seems to me they are nested under `data` field

Reply:
Thank you for the catch, forgot to change this